store/src/java-test/com/zimbra/cs/service/mail/SendMsgTest.java

```diff
@@ -255,4 +263,60 @@ public void sendUpload() throws Exception {
         MimeMessage mm = msg.getMimeMessage();
         Assert.assertEquals("correct top-level MIME type", "multipart/alternative", new ZContentType(mm.getContentType()).getBaseType());
     }
+
+    @Test
+    public void testSendMailEventLoggerTest() throws Exception {
+
+        EventLogHandler.Factory mockFactory = Mockito.mock(EventLogHandler.Factory.class);
+        EventLogHandler mockHandler = Mockito.mock(EventLogHandler.class);
+        Mockito.doReturn(mockHandler).when(mockFactory).createHandler(Mockito.anyString());
+
+        EventLogger.registerHandlerFactory("testhandler", mockFactory);
+
+        EventLogger.ConfigProvider mockConfigProvider = Mockito.mock(EventLogger.ConfigProvider.class);
+
+        Multimap<String, String> mockConfigMap = ArrayListMultimap.create();
+        mockConfigMap.put("testhandler", "");
+        Mockito.doReturn(mockConfigMap.asMap()).when(mockConfigProvider).getHandlerConfig();
+
+        Mockito.doReturn(1).when(mockConfigProvider).getNumThreads(); //ensures sequential event processing
+
+        EventLogger.getEventLogger(mockConfigProvider).startupEventNotifierExecutor();
+
+        Account acct = Provisioning.getInstance().getAccountByName("test@zimbra.com");
+        Mailbox mbox = MailboxManager.getInstance().getMailboxByAccount(acct);
+
+        // first, add draft message
+        MimeMessage mm = new MimeMessage(Session.getInstance(new Properties()));
+        mm.setRecipients(RecipientType.TO, "rcpt1@zimbra.com,rcpt2@zimbra.com");
+        mm.saveChanges();
+        ParsedMessage pm = new ParsedMessage(mm, false);
+        int draftId = mbox.saveDraft(null, pm, Mailbox.ID_AUTO_INCREMENT).getId();
+
+        // then send a message referencing the draft
+        Element request = new Element.JSONElement(MailConstants.SEND_MSG_REQUEST);
+        request.addElement(MailConstants.E_MSG).addAttribute(MailConstants.A_DRAFT_ID, draftId).addAttribute(MailConstants.A_SEND_FROM_DRAFT, true);
+        Element response = new SendMsg().handle(request, ServiceTestUtil.getRequestContext(acct));
+
+        ArgumentCaptor<Event> captor = ArgumentCaptor.forClass(Event.class);
+        Mockito.verify(mockHandler, Mockito.times(2)).log(captor.capture());
+
+        Assert.assertNotNull(captor.getAllValues());
+        Assert.assertEquals(2, captor.getAllValues().size());
+
+        Assert.assertEquals(Event.EventType.SENT, captor.getAllValues().get(0).getEventType());
+        Assert.assertEquals(Event.EventType.SENT, captor.getAllValues().get(1).getEventType());
+
+        Assert.assertEquals("test@zimbra.com", captor.getAllValues().get(0).getContextField(Event.EventContextField.SENDER));
+        Assert.assertEquals("test@zimbra.com", captor.getAllValues().get(1).getContextField(Event.EventContextField.SENDER));
+
+        Assert.assertEquals("rcpt1@zimbra.com", captor.getAllValues().get(0).getContextField(Event.EventContextField.RECEIVER));
```

Comment:
Just wondering if the ordering here is guaranteed. Will recipients always be logged in the same order as they are listed in the message header?

Reply:
They will be logged in the same order here because there is only 1 consumer thread:
`Mockito.doReturn(1).when(mockConfigProvider).getNumThreads()`